jina/serve/networking.py

```diff
@@ -89,15 +112,55 @@ async def remove_connection(self, address: str):
 
         return None
 
-    def get_next_connection(self):
+    def _create_connection(self, address):
+        try:
+            parsed_address = urlparse(address)
+            address = parsed_address.netloc if parsed_address.netloc else address
+            use_tls = parsed_address.scheme in TLS_PROTOCOL_SCHEMES
+        except:
+            use_tls = False
+
+        stubs, channel = GrpcConnectionPool.create_async_channel_stub(
+            address, tls=use_tls, summary=self.summary
+        )
+        return stubs, channel
+
+    async def _destroy_connection(self, connection, grace=0.5):
+        # we should handle graceful termination better, 0.5 is a rather random number here
+        await connection.close(grace)
+
+    async def get_next_connection(self):
         """
         Returns a connection from the list. Strategy is round robin
         :returns: A connection from the pool
         """
+        return await self._get_next_connection(num_retries=3)
+
+    async def _get_next_connection(self, num_retries=0):
+        """
+        :param num_retries: how many retries should be performed when all connections are currently unavailable
+        :returns: A connection from the pool
+        """
         try:
-            connection = self._connections[self._rr_counter]
+            connection = None
+            for i in range(len(self._connections)):
+                internal_rr_counter = (self._rr_counter + i) % len(self._connections)
+                connection = self._connections[internal_rr_counter]
+                # connection is None if it is currently being reset. In that case, try different connection
+                if connection is not None:
+                    break
+            if connection is None and num_retries <= 0:
+                # if still None, then all connections are currently being reset
+                if num_retries <= 0:
+                    raise ConnectionError(
+                        f'All connections in {self._connections} are currently being reset and cannot be used.'
+                    )
```

Comment:
Here it would probably be better to raise an Exception that inherits from `BaseJinaExceptions`, but there isn't one that fits (in particular, `InternalNetworkError` sounds right, but is reserved as a wrapper around `AioRpcError` and gets special treatment).
Creating a new Exception class just for this seemed excessive.

@JoanFM any suggestions?

Reply:
u can create a new one with no problem